### ISSUE TITLE
Training issue with `valid_mask` in multi-batch

### ISSUE THREAD

**Issue by Kitsunetic** [JarrentWu1031/AnimateAnyMesh#10]
When training my custom script (since no official training script is provided), I encounter nan losses when using a batch_size > 1 with the VAE model.

The problem seems related to varying numbers of vertices across samples. Specifically:
- The `valid_mask` has shape $\mathbb{R}^{B \times N \times N}$.
- For samples with fewer than $N$ vertices, some rows of `valid_mask` are entirely `False`.
- In these cases, the `DyMeshVAE` model produces nan loss.

Question:
- How do you handle multi-batch training in this setting?
- Is there a recommended approach for avoiding nan losses when the number of vertices differs between samples?

**Comment by JarrentWu1031** [comment]
Well, that is strange. In our training scheme, the adj_matrix is allowed to have "all-False" rows. You could do a simple check as follows: modify Line 110 of test_drive.py (https://github.com/JarrentWu1031/AnimateAnyMesh/blob/18416e3d5c2361a0a9bc91e527240bb091e8a332/test_drive.py#L110C1-L111C1) as:
```
opt.max_length = vertices.shape[0] + 512
```
Then, the test sample will be padded, which corresponds to the training script. And the adj_matrix would have "all-False" rows in its last 512 rows. You are supposed to get similar rendering results. For suggestions, I recommend you to write the training script according to the test_drive.py file, you could check whether the shape/dtype of each tensor match the corresponding ones in test_drive.py. I think you will find where the error lies in.

**Comment by Kitsunetic** [comment]
Hmm... the main reason might be due to different data setting with your training script.
However, I still think the VAE model produces nan result when there is all False adj_matrix.

<img width="671" height="817" alt="Image" src="https://github.com/user-attachments/assets/690cbcd1-1f99-43dc-ab60-879dfbaf524d" />

It is clear when there is all-False row, SDPA produces nan result in the specific rows.
Then, the nan values contaminate all other rows in here:
https://github.com/JarrentWu1031/AnimateAnyMesh/blob/18416e3d5c2361a0a9bc91e527240bb091e8a332/networks/dymeshvae.py#L214

I think attention mask is missing in this line. I will test with additional attention mask here.

**Comment by JarrentWu1031** [comment]
<img width="747" height="535" alt="Image" src="https://github.com/user-attachments/assets/f1849a46-79a6-45ae-8e16-ed0783f93b4b" />

Well, that is even tricker. As you can see, I have tried to replicate your setup, the output is still valid. Hope I do not miss anything.

About L214 of dymeshvae.py, only NeighborAttention needs attention mask to aggregate the adjacency information, while SyncAttention do not need attention mask. Therefore we drop the mask option for simplify. Anyway, thanks for your advice!

**Comment by JarrentWu1031** [comment]
Well, I think it is the version of Pytorch that causes the difference. The issue seems to be solved now: https://github.com/pytorch/pytorch/pull/157727.
In fact, there is a simple way to avoid the nan issue when computing sdpa, which is to switch the all-False row to all-True row. In our training script, we only calculate the error within the valid mask of each instance. Therefore the padded vertices would not influence the optimization.

**Comment by Kitsunetic** [comment]
Ah, I learn new thing! I also think the issue came from the different pytorch version.
Thank you, I will check about it!

**Comment by Kitsunetic** [comment]
Hi Jarrent!
I’m at ECCV 2026 in Malmö and was wondering if you’re attending too.
If so, would you be interested in a quick coffee chat during the conference?
I’d love to chat about mesh animation generation and your work.
No worries if you’re busy!

**Comment by JarrentWu1031** [comment]
> Hi Jarrent! I’m at ECCV 2026 in Malmö and was wondering if you’re attending too. If so, would you be interested in a quick coffee chat during the conference? I’d love to chat about mesh animation generation and your work. No worries if you’re busy!

Hi Kitsunetic! Glad to hear from you. Lately I've been busy and was not able to make it to attend ECCV26. Hope we can meet in the future! Anyway, wish you have a good time in Malmö~
